src/sqlfluff/core/parser/grammar/anyof.py

```diff
@@ -171,6 +172,13 @@ def match(
         matched_segments: MatchResult = MatchResult.from_empty()
         unmatched_segments: Tuple[BaseSegment, ...] = segments
         n_matches = 0
+
+        # Keep track of the number of times each option has been matched.
+        available_options, _ = self._prune_options(
+            segments, parse_context=parse_context
+        )
+        available_option_counter = {str(o): 0 for o in available_options}
```

Comment:
The change is pretty simple: Keep a counter of each time an available_option is used and then exit if we're adding an already used available_option.